src/VisualStudio/Core/Next/Remote/JsonRpcClient.cs

```diff
@@ -88,23 +93,76 @@ public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments,
             return Extensions.InvokeAsync(_rpc, targetName, arguments, funcWithDirectStreamAsync, _cancellationToken);
         }
 
+        public void Dispose()
+        {
+            OnDisposed();
+
+            _rpc.Dispose();
+        }
+
+        protected void StartListening()
+        {
+            // due to this issue - https://github.com/dotnet/roslyn/issues/16900#issuecomment-277378950
+            // _rpc need to be explicitly started
+            _rpc.StartListening();
+        }
+
+        protected virtual void OnDisposed()
+        {
+            // do nothing
+        }
+
+        // these are for debugging purpose. once we find out root cause of the issue
+        // we will remove these.
+        private static JsonRpcDisconnectedEventArgs s_debuggingLastDisconnectReason;
+        private static string s_debuggingLastDisconnectCallstack;
+
+        private JsonRpcDisconnectedEventArgs _debuggingLastDisconnectReason;
+        private string _debuggingLastDisconnectCallstack;
+
         private bool ReportUnlessCanceled(Exception ex, CancellationToken cancellationToken)
         {
             if (cancellationToken.IsCancellationRequested)
             {
                 return true;
             }
 
-            // save extra info using NFW
-            ReportExtraInfoAsNFW(ex);
+            s_debuggingLastDisconnectReason = _debuggingLastDisconnectReason;
+            s_debuggingLastDisconnectCallstack = _debuggingLastDisconnectCallstack;
 
-            // make it to explicitly crash to get better info
-            FatalError.Report(ex);
+            // send NFW to figure out why this is happening
+            ReportExtraInfoAsNFW(ex);
 
             GC.KeepAlive(_debuggingLastDisconnectReason);
             GC.KeepAlive(_debuggingLastDisconnectCallstack);
 
-            return Contract.FailWithReturn<bool>("shouldn't be able to reach here");
+            return true;
+        }
+
+        private static bool s_reported = false;
+
+        /// <summary>
+        /// Show info bar and throw its own cancellation exception until 
+        /// we figure out this issue
+        /// https://devdiv.visualstudio.com/DevDiv/_workitems/edit/453544
```

Comment:
❕ Link to bug on dotnet/roslyn or include text to explain the issue for a user that does not have access to this link.